### PR TITLE
[18] Make diagram element layouts uniform during creation

### PR DESCRIPTION
**Pull request by scosta-obeo**
Bug: https://github.com/eclipse-sirius/sirius-desktop/issues/18

### CODE REVIEW

**Review by lfasani — CHANGES_REQUESTED** [review]

File: plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/business/internal/view/SiriusLayoutDataManagerImpl.java [hunk 1]
```diff
@@ -620,30 +673,36 @@ private Command getCreatedViewsCommandFromLayoutType(List<IAdaptable> createdVie
...
         IFigure figure = part.getFigure();
         if (previousCenterLocation == null) {
             // Get center (reference point)
-            Point referencePoint = layoutHelper.getReferencePosition(host.getContentPane(), ((FigureCanvas) host.getViewer().getControl()).getViewport(), host);
+            Point referencePoint;
+            if (host instanceof DiagramEditPart) {
+                referencePoint = layoutHelper.getReferencePosition(host.getContentPane(), ((FigureCanvas) host.getViewer().getControl()).getViewport(), host);
+            } else {
+                referencePoint = new Point(20, 20);
+            }
```
[the remark this answers is not in the document]
Comment (scosta-obeo) [reply plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/business/internal/view/SiriusLayoutDataManagerImpl.java:689]: Done

File: plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/graphical/figures/SiriusLayoutHelper.java [hunk 2]
```diff
@@ -51,6 +54,62 @@ public SiriusLayoutHelper(IGraphicalEditPart editPart) {
...
+        if (clobber != null) {
+            theBounds.setLocation(updateClobberedPosition(clobber, null));
+            return validatePosition(parent, theBounds, exceptions);
+        }
+        return theBounds.getLocation();
+    }
+
+    /**
+     * Return a child figure (except figure in <code>exceptions</code>) that is already occupying the a position within
+     * the supplied <code>bounds</code> figure.
```
[the remark this answers is not in the document]
Comment (scosta-obeo) [reply plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/graphical/figures/SiriusLayoutHelper.java:81]: Done

File: plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/graphical/figures/SiriusLayoutHelper.java [hunk 3]
```diff
@@ -51,6 +54,62 @@ public SiriusLayoutHelper(IGraphicalEditPart editPart) {
...
+     * 
+     * @param parent
+     *            The parent figure of the newly added child.
+     * @param bounds
+     *            the area being tested.
+     * @return a child figure already occupying the supplied <code>bounds</code> or <code>null</code> if the bounds are
+     *         unoccupied or occupied by an <code>exceptions</code>-list figure.
+     */
+    public IFigure findFigureInExcept(IFigure parent, Rectangle bounds, Collection<IFigure> exceptions) {
+        ListIterator<? extends IFigure> listIterator = parent.getChildren().listIterator();
```
[the remark this answers is not in the document]
Comment (scosta-obeo) [reply plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/graphical/figures/SiriusLayoutHelper.java:91]: I prefer to be consistent with the other `findFigure` function. And as one figure is always inside another, I don't think InParent is necessary.

File: plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/internal/refresh/layout/SiriusCanonicalLayoutCommand.java [hunk 4]
```diff
@@ -72,41 +88,45 @@ public SiriusCanonicalLayoutCommand(TransactionalEditingDomain domain, DiagramEd
...
-            @Override
-            public void run() {
-                if (childViewsAdapters == null && centeredChildViewsAdapters == null && borderedChildViewsAdapters == null) {
-                    executeLayoutOnDiagramOpening();
-                } else {
-                    executeLayoutDueToExternalChanges();
-                }
+        Display.getDefault().asyncExec(() -> {
+            if (diferredExecution) {
+                // Defer execution to the end of the async execution queue.
```
[the remark this answers is not in the document]
Comment (scosta-obeo) [reply plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/internal/refresh/layout/SiriusCanonicalLayoutCommand.java:110]: No, not diferred, not differed but deferred

**Pull request merged**